### PR TITLE
dialects: (builtin) DenseIntOrFPElementsAttr: make data private and use getters instead

### PR DESCRIPTION
**Pull request by jorendumoulin**
To prepare for the change of internal data representation, this PR makes the actual data of the attribute a private attribute. Everthing is accessed with the getters `get_attrs` (IntegerAttr, ...) and `get_values` (int, float).

The check whether the dense is set by a scalar is pretty common, so i included an additional helper function for that as well

### CODE REVIEW

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -1875,6 +1875,26 @@ def tensor_from_list(
...
+        return tuple(el.value.data for el in self._data.data)
+
+    def get_attrs(self) -> Sequence[AnyIntegerAttr] | Sequence[AnyFloatAttr]:
+        """
+        Return all elements of the dense attribute in their relevant
+        attribute representation (IntegerAttr / FloatAttr)
+        """
+        return self._data.data
+
+    def is_scalar(self) -> bool:
```
Comment (math-fehr) [inline xdsl/dialects/builtin.py:1891]: Can we rename this to `is_splat`?
`is_scalar` would mean that there is a single element in the tensor, where here we want to say that all elements are the same.

**Review by math-fehr — APPROVED** [review]
Seems sensible to me, I would just rename `is_scalar` to `is_splat` though

File: xdsl/dialects/builtin.py (hunk 1 above)
Comment (jorendumoulin) [reply xdsl/dialects/builtin.py]: Okay, thanks!

File: xdsl/transforms/csl_stencil_bufferize.py [hunk 2]
```diff
@@ -370,9 +375,11 @@ def match_and_rewrite(self, op: arith.ConstantOp, rewriter: PatternRewriter, /):
             return
         assert isinstance(op.value, DenseIntOrFPElementsAttr)
         assert isa(op.value.type, TensorType[Attribute])
-        typ = DenseIntOrFPElementsAttr(
-            [tensor_to_memref_type(op.value.type), op.value.data]
+        memref_type = cast(
+            RankedStructure[AnyFloat | IntegerType | IndexType],
+            tensor_to_memref_type(op.value.type),
         )
+        typ = DenseIntOrFPElementsAttr.from_list(memref_type, op.value.get_values())
```
Comment (superlopuh) [inline xdsl/transforms/csl_stencil_bufferize.py:382]: why this change? it should still work, as-is, no?

File: xdsl/transforms/experimental/stencil_tensorize_z_dimension.py [hunk 3]
```diff
@@ -425,7 +428,15 @@ def match_and_rewrite(self, op: ConstantOp, rewriter: PatternRewriter, /):
...
-                        ConstantOp(DenseIntOrFPElementsAttr([typ, op.value.data]))
+                        ConstantOp(
+                            DenseIntOrFPElementsAttr.from_list(
+                                cast(
+                                    RankedStructure[AnyFloat | IntegerType | IndexType],
+                                    typ,
+                                ),
+                                op.value.get_values(),
+                            )
+                        )
```
Comment (superlopuh) [inline xdsl/transforms/experimental/stencil_tensorize_z_dimension.py:439]: same here, let's preserve this behaviour in this PR

File: xdsl/transforms/linalg_to_csl.py [hunk 4]
```diff
@@ -45,9 +45,9 @@ def get_scalar_const(op: SSAValue) -> AnyFloatAttr | AnyIntegerAttr | None:
         isinstance(op, OpResult)
         and isinstance(op.op, arith.ConstantOp)
         and isa(val := op.op.value, DenseIntOrFPElementsAttr)
-        and val.data.data.count(val.data.data[0]) == len(val.data.data)
+        and val.is_splat()
     ):
-        return val.data.data[0]
+        return val.get_attrs()[0]
```
Comment (superlopuh) [inline xdsl/transforms/linalg_to_csl.py:50]: very nice

File: xdsl/dialects/builtin.py [hunk 5]
```diff
@@ -1730,7 +1730,7 @@ class DenseIntOrFPElementsAttr(TypedAttribute, ContainerType[AnyDenseElement]):
         | RankedStructure[IndexType]
         | RankedStructure[AnyFloat]
     ]
-    data: ParameterDef[ArrayAttr[AnyIntegerAttr] | ArrayAttr[AnyFloatAttr]]
+    _data: ParameterDef[ArrayAttr[AnyIntegerAttr] | ArrayAttr[AnyFloatAttr]]
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:1733]: let's unprivate it now that all the uses are migrated?

File: xdsl/transforms/csl_stencil_bufferize.py (hunk 2 above)
Comment (jorendumoulin) [reply xdsl/transforms/csl_stencil_bufferize.py:382]: yes, but then you have to access the private _data.
I can unprivatize data and make exceptions for these more trivial cases

**Review by superlopuh — APPROVED** [review]
